extensions/community/PokiGamesSDK.json

```diff
@@ -0,0 +1,321 @@
+{
+  "author": "",
+  "category": "Ads",
+  "extensionNamespace": "",
+  "fullName": "Poki Games SDK",
+  "helpPath": "",
+  "iconUrl": "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA1NzguODggMjczLjE2Ij4KCTxwYXRoIGZpbGw9ImJsYWNrIiBkPSJNNTQxIDg1LjhjMjEgMCAzOC0xNy40IDM4LTM4LjcgMC0yMC40LTE3LjQtMzcuNy0zOC0zNy43UzUwMy41IDI2LjYgNTAzLjUgNDdjMCAyMS40IDE3IDM4LjggMzcuNyAzOC44eiI+Cgk8L3BhdGg+Cgk8cGF0aCBmaWxsPSJibGFjayIgZD0iTTUwNCA5OS4zdjExMC41TDQ3MCAxMzZjMTUuNC0xOS4zIDIxLjctNDYuNCAyMi40LTY2VjU0aC02MC4ydjE1LjVjLTEgMTkuNi03LjMgMjguOC0yNC44IDM0LjVWMEgzNDZ2NDQuNEMzMjQgMjEuNiAyOTMgOS42IDI1Ni43IDkuNmMtNDAgMC03MyAxMy44LTk1LjYgNDBhMTE4IDExOCAwIDAgMC0xNyAyNi4yQzEyOSA2MC42IDEwOC40IDU0IDg0IDU0SDB2MjE5LjJoNTZWMjIxaDI4YzI0LjQgMCA0NS04LjMgNjAtMjMuN2ExMTYuMyAxMTYuMyAwIDAgMCAxNy42IDI3YzIyLjcgMjUuOCA1NS41IDM5LjMgOTUgMzkuMyAzOC42IDAgNzEtMTMuNSA5My42LTM5bDQuNS01LjZoNTIuN3YtNTQuM2w4LTIgMjYgNTYuNGgxMzYuOFY5OS40aC03NHptLTQxOCA2NEg1NnYtNTNoMjkuMmM3LjYgMCAyMC4zIDUuMiAyMC4zIDI1LjUgMCAxMy01LjQgMjcuMy0xOS40IDI3LjN6bTE3MCA2MGMtNDcuMyAwLTg1LjctMzkuNC04NS43LTg3LjhDMTcwLjMgODkgMjA5LjUgNTAgMjU2IDUwczg1LjYgMzkgODUuNiA4NS41YzAgNDguNC0zOC40IDg3LjgtODUuNiA4Ny44eiI+Cgk8L3BhdGg+Cgk8cGF0aCBmaWxsPSJibGFjayIgZD0iTTI1NiAxNDJjLTM4IDAtNTcgNi41LTY1IDguNiA2LjcgMzAuNSAzMy4yIDUzLjMgNjUgNTMuM3M1OC4zLTIzIDY1LTUzLjRjLTgtMi0yNy4yLTguNy02NS04Ljd6Ij4KCTwvcGF0aD4KPC9zdmc+Cg==",
+  "name": "PokiGamesSDK",
+  "previewIconUrl": "",
+  "shortDescription": "Extension that implements Poki SDK so that your game can be hosted on Poki.",
+  "version": "0.0.1",
+  "description": [
+    "Please refer to [Poki's guide](https://sdk.poki.com/sdk-documentation.html) to know when to call the different actions.",
+    "",
+    "To initialize the SDK, there is an asynchronous action \"Load SDK\" that is available."
+  ],
+  "tags": [
+    "poki",
+    "ads",
+    "sdk"
+  ],
+  "authorIds": [
+    "tgIIofwSWVNtHsFW7AR2imu8gCj1"
+  ],
+  "dependencies": [],
+  "eventsFunctions": [
+    {
+      "fullName": "",
+      "functionType": "Action",
+      "name": "onFirstSceneLoaded",
+      "sentence": "",
+      "events": [
+        {
+          "type": "BuiltinCommonInstructions::JsCode",
+          "inlineCode": "gdjs._pokiGamesSdkExtension = {};",
+          "parameterObjects": "",
+          "useStrict": true,
+          "eventsSheetExpanded": false
+        }
+      ],
+      "parameters": [],
+      "objectGroups": []
+    },
+    {
+      "async": true,
+      "description": "Load Poki SDK.",
+      "fullName": "Load SDK",
+      "functionType": "Action",
+      "group": "Poki Games",
+      "name": "LoadSDK",
+      "sentence": "Load Poki SDK",
```

Comment:
We ended up think that it shouldn't be done in the OnFirstSceneLoaded callback to leave the possibility to the user to easily disable the SDK if they want to export the game for a platform that is not Poki. Maybe this logic should be implemented in other platform SDK implementation extensions

Reply:
What happens if players have an intermittent internet connection and the action fails?
Can the SDK be downloaded as soon as the internet connection comes back?
How do creators know when to try again?
What happens if this action is ran in loop every frame? Is there some kind of protection against flooding?

Reply:
Do you think it's worth adding those checks although this SDK is only useful when the game is played on poki platform that is a website?

Reply:
Maybe not

Reply:
> We ended up think that it shouldn't be done in the OnFirstSceneLoaded callback to leave the possibility to the user to easily disable the SDK if they want to export the game for a platform that is not Poki

I disagree with this, such an sdk should likely be a no-op anyways when it is not ran on poki. If poki's SDK crashes the game or something when it is ran outside poki, then you should add code that checks if the game is running on poki based for example on the URL before loading the SDK. 

Asking the user to themselves detect if the game is running on poki is just unnecessary extra work. Not to mention, it is terrible UX to have other actions just no-op by default (or crash the game) if the user didn't know to call an "initialize" action beforehand.

Reply:
The game should be working the same on most platforms. The SDKs only provide augmentations, not a different game: e.g. on newgrounds, when a specific action is done you'll unlock a medal, with Poki when losing you'll trigger an ad... Those augmentations are not new game logic, they are just a piece of additional platform-specific logic to trigger during the normal game logic. There is no need to separate codebases! Just call the run ad action when you want an ad to run, unlock medal when you want to unlock a medal... And the extension can simply call the SDK.

If the game is not running in the platform, it should still try if possible to do what it's told: e.g. a newgrounds extensions can still unlock a medal even if you are not running the game from newgrounds, it just opens a login popup instead of auto-logging-in. If the SDK won't allow using it on other sites, e.g. with poki ads, then just create a non-breaking no-op default behavior. The SDK gives the platform a hint to display an ad, if there is no platform then just don't send any hint, don't force the game to run exclusively on that platform!

Reply:
Users could write an extension to abstract the platform-dedicated extensions they use from their games?